File: packages/typescript/src/surfaces/runtime.ts
import type { YaiEnvelope } from '../envelope';
import { YAI_COMPAT_OPERATIONS } from '../operations';
import {
  type RuntimeActiveCaseReason,
  type RuntimeAuthPosture,
  type RuntimeCasePosture,
  type RuntimeClientPosture,
  type RuntimeControlActionAvailability,
  type RuntimeControlActionReason,
  type RuntimeControlPlan,
  type RuntimeControlPlanAction,
  type RuntimeControlPlanServiceManager,
  type RuntimeLifecycleState,
  type RuntimeOperatorContextPosture,
  type RuntimeSealReason,
  type RuntimeSealedPosture,
  type RuntimeServiceControlPosture,
  type RuntimeServiceInstallPosture,
  type RuntimeServiceManagerPosture,
  type RuntimeServiceMode,
  type RuntimeOperationalReadiness,
  type RuntimeSessionPosture,
  type RuntimeServiceReadiness,
  type RuntimeServiceStatus,
  SystemSurface
} from './system';

export type {
  RuntimeActiveCaseReason,
  RuntimeAuthPosture,
  RuntimeCasePosture,
  RuntimeClientPosture,
  RuntimeControlActionAvailability,
  RuntimeControlActionReason,
  RuntimeControlPlan,
  RuntimeControlPlanAction,
  RuntimeControlPlanServiceManager,
  RuntimeLifecycleState,
  RuntimeOperatorContextPosture,
  RuntimeSealReason,
  RuntimeSealedPosture,
  RuntimeServiceControlPosture,
  RuntimeServiceInstallPosture,
  RuntimeServiceManagerPosture,
  RuntimeServiceMode,
  RuntimeOperationalReadiness,
  RuntimeSessionPosture,
  RuntimeServiceReadiness,
  RuntimeServiceStatus
};

export interface RuntimeControlPlanStatus extends RuntimeServiceStatus {
  controlPlan?: RuntimeControlPlan;
}

// Deprecated compatibility alias; prefer SystemSurface and client.system.
export class RuntimeSurface extends SystemSurface {
  lifecycle(request?: unknown): Promise<YaiEnvelope<RuntimeServiceStatus>> {
    return this.invoke<RuntimeServiceStatus>(
      YAI_COMPAT_OPERATIONS.runtimeServiceLifecycleInspect,
      request,
      'Runtime service transport not configured; lifecycle status is unavailable until a real transport is attached'
    );
  }

  health(request?: unknown): Promise<YaiEnvelope<RuntimeServiceStatus>> {
    return this.invoke<RuntimeServiceStatus>(
      YAI_COMPAT_OPERATIONS.runtimeServiceHealthInspect,
      request,
      'Runtime service transport not configured; health status is unavailable until a real transport is attached'
    );
  }

  controlPlan(): Promise<YaiEnvelope<RuntimeControlPlanStatus>> {
    return this.invoke<RuntimeControlPlanStatus>(
      YAI_COMPAT_OPERATIONS.runtimeServiceControlPlan,
      { mode: 'plan-only' },
      'Runtime service transport not configured; control planning is unavailable until a real transport is attached'
    );
  }
}

File: packages/rust/src/surfaces/runtime.rs
use crate::envelope::YaiEnvelope;
use crate::operations::{YAI_COMPAT_OPERATIONS, YAI_OPERATIONS};
use crate::status::YaiResult;
use crate::surfaces::system::RuntimeServiceStatus;
use crate::transport::YaiTransport;

pub struct RuntimeSurface<TTransport: YaiTransport> {
    transport: TTransport,
}

impl<TTransport: YaiTransport> RuntimeSurface<TTransport> {
    pub fn new(transport: TTransport) -> Self {
        Self { transport }
    }

    pub fn status_inspect(&self) -> YaiResult<YaiEnvelope<RuntimeServiceStatus>> {
        self.transport
            .invoke::<RuntimeServiceStatus>(YAI_OPERATIONS.system_status, None)
    }

    pub fn control_plan_inspect(&self) -> YaiResult<YaiEnvelope<RuntimeServiceStatus>> {
        self.transport.invoke::<RuntimeServiceStatus>(
            YAI_COMPAT_OPERATIONS.runtime_service_control_plan,
            Some(serde_json::json!({ "mode": "plan-only" })),
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::{NotConfiguredTransport, YaiClient, YaiError};

    #[test]
    fn runtime_status_not_configured_is_not_success() {
        let client = YaiClient::new(NotConfiguredTransport);
        let result = client.runtime().status_inspect();
        assert!(matches!(
            result,
            Err(YaiError::TransportNotConfigured {
                operation_id: "system.status"
            })
        ));
    }

    #[test]
    fn runtime_control_plan_not_configured_is_not_success() {
        let client = YaiClient::new(NotConfiguredTransport);
        let result = client.runtime().control_plan_inspect();
        assert!(matches!(
            result,
            Err(YaiError::TransportNotConfigured {
                operation_id: "runtime.service.control.plan"
            })
        ));
    }
}
